### DISCUSSION TITLE
Proposal: Idiomatic Code Refactoring using Pipeline Symbol in LSP

### DISCUSSION THREAD

**Discussion by katsmil** [gleam-lang/gleam#2573] [Ideas & suggestions]
I have been working on a potential improvement for the Gleam project, focusing on extending the Language Server Protocol (LSP) implementation to provide additional refactoring hints. Specifically, this extension introduces idiomatic code refactoring using the pipeline symbol (|>), aimed at helping users write more idiomatic code.

### Visual Representation
To make the proposed language refactor more concrete, I've created a GIF showcasing how the suggested refactor can be executed by the editor. In this example, the CodeAction interface of the Language Server Protocol is utilized. That means only the lightbulb indicates a hint, there are no squiggly lines.

In this demo, all parts are converted into a single piped expression. However, it might be important for users to choose which refactor to execute. This could include:

- Convert all parts into a single piped expression.
- Using the pipeline symbol on specific lines rather than converting all parts into a single expression.
- Choosing to inline a variable.

![gif_pipeline_gleam](https://github.com/gleam-lang/gleam/assets/29780165/da78dd6d-5306-44cd-9c94-1e581b2011f6)

### Diagnostic vs. Code Action:

Now, if we agree on integrating this addition into the LSP for Gleam, how should we present the suggestion? There are two potential approaches:

- Diagnostic Approach (Squiggly Lines): This involves highlighting the areas of improvement with squiggly lines in the code editor, for example a blue line for hints.
- Code Action Approach (Lightbulb): Alternatively, we can present the suggestion solely as a code action.

I would love to hear your thoughts and preferences on this matter.

**Comment by giacomocavalieri** [comment]
That's an interesting refactor! I'd go for the code action approach for sure, sometimes people might want to give names to intermediate steps and seeing the squiggly line would be distracting

**Reply by lpil** [reply]
I agree! Both approaches are valid, so don't want to annoy folks using one over the other.

This would make a great code action ✨
